esp-hal/src/work_queue.rs

```diff
@@ -0,0 +1,451 @@
+    ///
+    /// The function returns whether the work item was accepted. If there is no driver currently
+    /// processing the queue, this function will return false to prevent removing the work item
+    /// from the queue.
+    ///
+    /// This function should be as short as possible.
+    pub(crate) post: fn(*const (), &mut T) -> bool,
+
+    /// Polls the status of a particular work item.
+    ///
+    /// Return `None` if the work item is not currently being processed.
+    ///
+    /// This function should be as short as possible.
+    pub(crate) poll: fn(*const (), &mut T) -> Option<Poll>,
+
+    /// Attempts to abort processing a work item.
+    ///
+    /// This function should be as short as possible.
+    pub(crate) cancel: fn(*const (), &mut T),
+
+    /// Called when the work queue becomes empty.
+    ///
+    /// This function should be as short as possible.
+    pub(crate) on_empty: fn(*const ()),
+}
+
+impl<T: Sync> VTable<T> {
+    pub(crate) const fn noop() -> Self {
+        Self {
+            post: |_, _| false,
+            poll: |_, _| None,
+            cancel: |_, _| (),
+            on_empty: |_| (),
+        }
+    }
+}
+
+struct Inner<T: Sync> {
+    head: *mut WorkItem<T>,
+    tail: *mut WorkItem<T>,
+    current: *mut WorkItem<T>,
+
+    data: *const (),
+    vtable: VTable<T>,
+}
+
+impl<T: Sync> Inner<T> {
+    /// Places a work item at the end of the queue.
+    fn enqueue(&mut self, ptr: *mut WorkItem<T>) {
+        if self.tail.is_null() {
+            // Queue is empty, set both `head` and `tail`.
+            self.tail = ptr;
+            self.head = ptr;
+        } else {
+            unsafe {
+                // Safety: we just checked that `tail` is not null.
+                (*self.tail).next = ptr;
+            }
+        }
+    }
+
+    /// Places a work item at the front of the queue.
+    fn enqueue_front(&mut self, ptr: *mut WorkItem<T>) {
+        unsafe { (*ptr).next = self.head };
+        self.head = ptr;
+        if self.tail.is_null() {
+            self.tail = ptr;
+        }
+    }
+
+    /// Pops and returns a work item from the start of the queue.
+    fn dequeue(&mut self) -> Option<*mut WorkItem<T>> {
+        let ptr = self.head;
+        // If the `head` is null, the queue is empty. Return None and do nothing.
+        if ptr.is_null() {
+            return None;
+        }
+
+        unsafe {
+            // Safety: we just checked that `ptr` is not null.
+            self.head = (*ptr).next;
+        }
+
+        // If the new `head` is null, the queue is empty. Clear the `tail` pointer.
+        if self.head.is_null() {
+            self.tail = core::ptr::null_mut();
+        }
+
+        Some(ptr)
+    }
+
+    /// Removes the item from the queue.
+    ///
+    /// Returns `true` if the work item was successfully removed, `false` if the work item was not
+    /// found in the queue.
+    fn remove(&mut self, ptr: *mut WorkItem<T>) -> bool {
```

Comment:
Rip, if not for this function, I would've suggested something lock-free

Reply:
Ah yeah

Also, given this would degrade to critical sections on some of the chips, I doubt it's worth the trouble of implementing/reviewing.